lib/Dialect/FIRRTL/FIRRTLReductions.cpp

```diff
@@ -1009,11 +1041,39 @@ struct EagerInliner : public OpReduction<firrtl::InstanceOp> {
     return success();
   }
 
+  /// Check if inlining the referenced module into the parent module would cause
+  /// inner symbol collisions.
+  bool hasInnerSymbolCollisions(FModuleOp referencedModule,
+                                FModuleOp parentModule) {
+    // Get the inner symbol tables for both modules
+    auto &targetTable = innerSymTables->getInnerSymbolTable(referencedModule);
+    auto &parentTable = innerSymTables->getInnerSymbolTable(parentModule);
+
+    // Check if any inner symbol name in the target module already exists
+    // in the parent module. Return failure() if a collision is found to stop
+    // the walk early.
+    LogicalResult walkResult = targetTable.walkSymbols(
+        [&](StringAttr name,
+            const hw::InnerSymTarget &target) -> LogicalResult {
+          // Check if this symbol name exists in the parent module
+          if (parentTable.lookup(name)) {
+            // Collision found, return failure to stop the walk
+            return failure();
+          }
+          return success();
+        });
+
+    // If the walk failed, it means we found a collision
+    return failed(walkResult);
+  }
```

Comment:
I kind of like this style of eagerly exit if the reduction pattern isn't trivially applicable. This is in contrast with trying to do a bunch of work to clean this up. I think this makes sense because the inner symbols will get dropped with other reductions.

Reply:
Totally agreed! It's also very annoying if the pattern applies too eagerly and then leads to invalid IR in a lot of cases. That just sends the reducer into busy work rabbit holes that don't really improve the reduction. :disappointed: